jmclient/jmclient/wallet_utils.py

```diff
@@ -1252,7 +1260,8 @@ def create_wallet(path, password, max_mixdepth, wallet_cls, **kwargs):
 
 
 def open_test_wallet_maybe(path, seed, max_mixdepth,
-                           test_wallet_cls=SegwitLegacyWallet, wallet_password_stdin=False, **kwargs):
+                           test_wallet_cls=SegwitLegacyWallet, wallet_password_stdin=False,
+                           changing_passphrase=False, **kwargs):
```

Comment:
So while this is viable (I think), I don't think this is how we should do it.

Changing passphrase isn't intrinsically associated with opening the wallet (which will always require the current passphrase).
As an example of how it could work, look at the code path for the `freeze` method. First we open the wallet, then we run `wallet_freezeutxo` which has a display callback function. The arguments and return values for that function are documented in detail, so you can take a look. That allows either Qt or CLI to update the state of the wallet, and save it, in response to user decisions. Obviously this passphrase case is simpler than that, but it can similarly be treated as a user action after the wallet load.

Let me know if that makes sense (notice how it wouldn't involve this extra argument passing-through).

Reply:
If I am right about what I wrote at the end there, then that can just be removed and everything else looks right apart from the `cli_` callback thing I mentioned.

Reply:
@AdamISZ thanks for the review, I'm currently digesting your feedback. I agree there's definitely room for improvement here.

> There's no reason to include the argument in open_test_wallet_maybe is there?

I think looking at commit 14d1f9e713d0cc3df3a7c2a9038849b6d0a46c7e on its own would help clarify the confusion.

This extra argument has no relation to the wallet_change_passphrase method. The job of the wallet_change_passphrase method is to ask for the *new* passphrase, and change it. It does *not* take care of asking for the *old* passphrase.

The big picture is the following. In the case that we are changing the passphrase, opening the wallet == asking for the *old* passphrase.

Thus, the extra argument is there because at the time we open the wallet, I want to add `old` to the prompt to decrypt the wallet. Instead of the usual "Enter wallet decryption passphrase: " prompt shown at the time of opening, I want to say "Enter old wallet decryption passphrase: ".

I'm hoping this clarifies things, but let me know if we are not on the same page.

Reply:
> Thus, the extra argument is there because at the time we open the wallet, I want to add old to the prompt to decrypt the wallet.

Right, that did cross my mind, thanks for the explanation.

It's interesting to consider the Qt/GUI case. There, `changepassphrase` will not be a method passed in at the same time you start a script(wallet-tool), but will be a menu item. There the two things are disconnected, so you have to open the wallet first, using the existing passphrase, before you even get a chance to request a particular action. In CLI we fold open the wallet and execute a method together in one step (from user perspective), so this question arises.

I think you should just not add that extra clutter to the open wallet workflow (which is already a bit more complicated than I'd like) and live with the fact that a user might mistakenly think they should enter the new passphrase at that time. In a sense this is not your problem as author of this PR - the problem is that we don't have a more high quality CLI interface, which should probably be interactive, perhaps an RPC based system to a daemon.

Suggestion: change "Enter wallet decryption passphrase:" to "Enter passphrase to decrypt wallet:" - this way a user is less likely to think they should be entering a *new* passphrase (and of course if they do, there is no harm done, and if they still don't get it after that, that's a bit dumb, especially considering they have doubtless opened their wallet this way multiple times before ...).

Reply:
For consistency, I suggest we also change:
`Enter wallet file encryption passphrase`
to
`[Re]enter passphrase to encrypt wallet`

Reply:
I wonder if that language is ideal to users when generating a wallet, but sure. I'm fine with either language.